server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java

```diff
@@ -904,4 +905,35 @@ public void testDisabledByAutoExpandReplicas() {
 
         assertThat(clusterState.getRoutingNodes().shardsWithState(UNASSIGNED), empty());
     }
+
+    public void testNodesWithoutAttributeAreIgnored() {
+        final Settings settings = Settings.builder()
+                .put(AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.getKey(), "zone")
+                .build();
+
+        final AllocationService strategy = createAllocationService(settings);
+
+        final Metadata metadata = Metadata.builder()
+                .put(IndexMetadata.builder("test").settings(settings(Version.CURRENT)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+                        .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 2)))
+                .build();
+
+        final ClusterState clusterState = applyStartedShardsUntilNoChange(
+                ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.get(Settings.EMPTY))
+                        .metadata(metadata)
+                        .routingTable(RoutingTable.builder()
+                                .addAsNew(metadata.index("test"))
+                                .build())
+                        .nodes(DiscoveryNodes.builder()
+                                .add(newNode("A-0", singletonMap("zone", "a")))
+                                .add(newNode("A-1", singletonMap("zone", "a")))
+                                .add(newNode("B-0", singletonMap("zone", "b")))
+                                .add(newNode("B-1", singletonMap("zone", "b")))
+                                .add(newNode("X-0", emptyMap()))
+                        ).build(), strategy);
+
+        assertThat(clusterState.getRoutingNodes().shardsWithState(UNASSIGNED), empty());
```

Comment:
where do we assert that `X-0` is ignored, i.e. does not hold a shard copy?

Reply:
Added in d7216bb7f0223d1cc92ee105f5e6137520c8fd3c